universal-hash/src/lib.rs

```diff
@@ -0,0 +1,154 @@
+//! Traits for [Universal Hash Functions].
+//!
+//! Universal hash functions select from a "universal family" of possible
+//! hash functions selected by a key. They are well suited to the purpose
+//! of "one time authenticators" for a sequence of bytestring inputs,
+//! as their construction has a number of desirable properties such as
+//! pairwise independence as well as amenability to efficient implementations,
+//! particularly when implemented using SIMD instructions.
+//!
+//! When combined with a cipher, such as in Galois/Counter Mode or the
+//! Salsa20 family AEAD constructions, they can provide the core functionality
+//! for a Message Authentication Code (MAC).
+//!
+//! [Universal Hash Functions]: https://en.wikipedia.org/wiki/Universal_hashing
+
+#![no_std]
+#![doc(html_logo_url = "https://raw.githubusercontent.com/RustCrypto/meta/master/logo_small.png")]
+#![warn(missing_docs, rust_2018_idioms)]
+
+#[cfg(feature = "std")]
+extern crate std;
+
+use generic_array::{ArrayLength, GenericArray};
+use generic_array::typenum::Unsigned;
+use subtle::{Choice, ConstantTimeEq};
+
+/// The `UniversalHash` trait defines a generic interface for universal hash
+/// functions.
+pub trait UniversalHash: Clone {
+    /// Size of a block (e.g. field element) this universal hash operates on
+    type BlockSize: ArrayLength<u8>;
+
+    /// Instantiate a universal hash function with the given key
+    fn new(key: GenericArray<u8, Self::BlockSize>) -> Self;
```

Comment:
So what about using `&GenericArray<u8, Self::BlockSize>` here, as well as in `update_block` and `verify`?

Reply:
Done, however I changed `verify` to:

```rust
fn verify(self, other: impl Into<Output<Self::BlockSize>>) -> Result<(), Error>
```

A `&GenericArray<u8, Self::BlockSize>` will still work, because I also added an `Into` impl for it.

Reply:
Ah, I wrote the other comment before seeing your reply. Let's discuss it there.